### PR TITLE
Use thread local storage for winbind context

### LINKED ISSUE #46: Multi-threaded libwbclient API should be supported
We use gssntlmssp library with support of winbind external server from multi-threaded application.
We see sporadic unexpected authentication failures during stress-test (when few requests are done in parallel) while everything is ok for non-parallel flow. Root-cause of the problem is unexpected interleaved data received by winbind, in this case request is dropped. We are getting an error for wbcAuthenticateUserEx call:

` wbc_status = wbcAuthenticateUserEx(&wbc_params, &wbc_info, &wbc_err);`

 For details of this call from gssntlmssp library,  see winbind_srv_auth function here:
 https://github.com/gssapi/gss-ntlmssp/blob/main/src/winbind.c

From winbind side:

```
[2021/01/06 08:29:19.646294, 10, pid=256853, effective(0, 0), real(0, 0), class=winbind] ../source3/winbindd/winbindd.c:763(process_request_send)
  process_request_send: process_request: request fn NTLMAUTH
...
[2021/01/06 08:29:19.646523,  0, pid=256853, effective(0, 0), real(0, 0), class=winbind] ../source3/winbindd/winbindd.c:1005(winbind_client_activity)
  winbind_client_activity[256138:PAM_AUTH_CRAP]:unexpected data from client - removing client
[2021/01/06 08:29:19.646616,  1, pid=256853, effective(0, 0), real(0, 0), class=winbind] ../source3/winbindd/winbindd_dual.c:337(wb_child_request_cleanup)
  wb_child_request_cleanup: keep orphaned subreq[0x55564bdc1b00]
```

We have got this feedback from Samba developers (Volker Lendecke):
_... one guess would be that the code using the gss-ntlmssp library is
multi-threaded. While the gss-ntlmssp library possibly is thread-safe
in general, its use of libwbclient is definitely not. Directly using
wbcAuthenticateUserEx() and other needs to be protected by a mutex, or
alternatively the library must create a wbcContext using wbcCtxCreate
in thread-local storage and then call wbcCtxAuthenticateUserEx(). The
wbcCtx*() calls are designed to be callable in a multi-threaded
environment, the wbcAuthenticateUserEx call is definitely not._

We need to have an option for using multi-threaded API of libwbclient from gssntlmssp.
One can use some compile-time option to specify the intended API or optionally we can just refactor the code to always use MT-safe calls.

gssntlmssp has few other winbind client calls - wbcInterfaceDetails and wbcCredentialCache - see src/winbind.c of gssntlmssp.
We haven't seen such races with them yet, but probably that's due very short time of request handling (local requests, no need to talk with DC). Probably all libwbclient calls should use the same approach.

### PR DESCRIPTION
**Pull request by simo5**
This allow multiple calls into winbindd from parallel threads without
suffering from a common mutex that serializaes all authentications.

The price to pay is that a new context is allocated for each thread, and
the way windbind works, this probably mean a separate file descriptor is
created for each of this thread.

The users of gssntlmssp in a multi-threaded environemnt will need to be
carufl to limit the number of threads they allow to use gssapi in this
case.

Resolves #46

### CODE REVIEW

**Review by kvv81 — APPROVED** [review]
It looks like exactly what we need, thanks!

File: src/winbind.c [hunk 1]
```diff
@@ -9,13 +9,75 @@
 
 #include <wbclient.h>
 
+#ifdef HAVE_PTHREAD
+#include <pthread.h>
+
+static int key_initialized = 0;
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/winbind.c:15]: ha, no, left over of a code iteration, I'll remove

**Comment by simo5** [comment]
@kvv81 any chance you can test if this work as expected?
Unfortunately the test suite does not have winbindd coverage yet...

**Comment by kvv81** [comment]
Hello,

Yes, I can test for regression with winbind API (checking that nothing is broken) soon. But full test - checking that NTLM authentication can really run in parallel (instead of serialization) will need hardware resources I don't have right now, this can take a week or so.

**Comment by kvv81** [comment]
@simo5 I found that your fix was done over old version of code we can't run in our environment.
Our code depends on some fixes done later in gssntlmssp, so I have just back-ported your patch to recent gss-ntlmssp:main code (it was one place - in winbind_get_creds - where 'patch' utility failed to apply hunk so I've done this manually).

According to my manual testing (few sequential auth by different users), no regression found.
I.e. NTLM authentication works as expected - as before.

Also after authentications I see that some new sockets are opened by process.
Output is expected since we keep winbind context per each thread who made request to winbind.
```
-lrwx------. 1 root   64 jan 22 09:44 148 -> socket:[1544612]
+lrwx------. 1 root   64 jan 22 09:45 152 -> socket:[1543036]
+lrwx------. 1 root   64 jan 22 09:45 153 -> socket:[1547934]
+lrwx------. 1 root   64 jan 22 09:45 154 -> socket:[1550322]
```

File: src/winbind.c [hunk 2]
```diff
@@ -9,13 +9,74 @@
 
 #include <wbclient.h>
 
+#ifdef HAVE_PTHREAD
+#include <pthread.h>
+
+static pthread_key_t key;
+static pthread_once_t key_once = PTHREAD_ONCE_INIT;
+
+/* not that the destructor is given directly the content of the key in
```
Comment (frozencemetery) [inline src/winbind.c:18]: typo - "note"

File: src/winbind.c [hunk 3]
```diff
@@ -9,13 +9,74 @@
...
+ * ptr, not the key itself. */
+static void key_destructor(void *ptr)
+{
+    wbcCtxFree((struct wbcContext *)ptr);
+}
+
+static void key_create(void)
+{
+    /* we have no way to report errors ... */
+    (void)pthread_key_create(&key, key_destructor);
```
Comment (frozencemetery) [inline src/winbind.c:28]: Shouldn't we abort here?  Failure will garbage up the context value; if we're lucky, the pthread_setspecific() call will fail with EINVAL:

       If an implementation detects that the value specified by the key argu‐
       ment to pthread_setspecific() does not refer to a a key value obtained
       from pthread_key_create() or refers to a key that has been deleted
       with pthread_key_delete(), it is recommended that the function should
       fail and report an [EINVAL] error.

**Review by frozencemetery — COMMENTED** [review]
Typo in commit message - "careful"

File: src/winbind.c (hunk 3 above)
Comment (simo5) [reply src/winbind.c:28]: I rather just have gssntlmp fail to operate than abort a user program

**Comment by simo5** [comment]
> @simo5 I found that your fix was done over old version of code we can't run in our environment.
> Our code depends on some fixes done later in gssntlmssp, so I have just back-ported your patch to recent gss-ntlmssp:main code (it was one place - in winbind_get_creds - where 'patch' utility failed to apply hunk so I've done this manually).

I'll fix this before merging, sorry for the annoyance.
 
> According to my manual testing (few sequential auth by different users), no regression found.
> I.e. NTLM authentication works as expected - as before.

Excellent!

> Also after authentications I see that some new sockets are opened by process.
> Output is expected since we keep winbind context per each thread who made request to winbind.
> 
> ```
> -lrwx------. 1 root   64 jan 22 09:44 148 -> socket:[1544612]
> +lrwx------. 1 root   64 jan 22 09:45 152 -> socket:[1543036]
> +lrwx------. 1 root   64 jan 22 09:45 153 -> socket:[1547934]
> +lrwx------. 1 root   64 jan 22 09:45 154 -> socket:[1550322]
> ```

I am planning to add another patch to make this behavior opt-in. (perhaps via env var, but I am open to suggestions). And instead create a new socket for each context in the normal case.
The reason is that often gssntlmsp is invoked behind several layers of abstraction (eg openldap libs -> sasl -> gssapi -> gssntlmssp, and therefore pretty unlikely general applications will control this properly.
OTOH specialized software that really need multithreaded performance, can easily set an environment variable (or something else?) to enable the more performant, but file descriptor consuming mode.

@kvv81  Let me know what you think about this approach.

File: src/winbind.c (hunk 2 above)
Comment (simo5) [reply src/winbind.c:18]: good catch

**Comment by simo5** [comment]
Rebased on main and fixed typos

**Comment by simo5** [comment]
New commit.

So here I implemented my idea to keep socket usage down for normal programs while allowing high performance programs to use thread local storage.

@kvv81 let me know if this is okay for your use case.

@frozencemetery I'd like also your opinion on this one.

File: src/winbind.c [hunk 4]
```diff
@@ -49,28 +49,54 @@ static struct wbcContext *winbind_get_context(void)
...
+#endif /* HAVE_PTHREAD */
+
+void *winbind_get_context(void)
+{
+#ifdef HAVE_PTHREAD
+    const char *envvar;
+
+    /* we return signal_ctx if we are told to use per thread
+     * context, the code knows how to cope correctly */
+    envvar = getenv("GSSNTLMSSP_WB_TLS_CTX");
```
Comment (kvv81) [inline src/winbind.c:63]: Can we implement also an option for compile-time switch of MT behaviour?

We have seen problems with passing environment variables for 'sudo/setcap' processes (privileged processes can suppress intrusion from non-privileged caller). So in our case we possibly prefer to apply some more general solution - like a specific macro during compilation.

File: src/winbind.c [hunk 5]
```diff
@@ -49,28 +49,54 @@ static struct wbcContext *winbind_get_context(void)
     return ctx;
 }
 
-#else /* HAVE_PTHREAD */
+static long signal_ctx = 0;
```
Comment (kvv81) [inline src/winbind.c:52]: Could you pls add a comment here explaining the usage of signal_ctx ?
I.e. that address of this variable is used as a marker/indicator of single-threaded winbind context.

Can't we use 'NULL' here as a marker (instead of &signal_ctx )?

File: src/winbind.c (hunk 4 above)
Comment (simo5) [reply src/winbind.c:63]: yes, I think we can do a compile time option

File: src/winbind.c (hunk 5 above)
Comment (simo5) [reply src/winbind.c:52]: As I was working at some point I need to differentiate between unallocated (NULL) and thread level context (&signal_ctx). Lemme see if I can simplify this one.

Comment (simo5) [reply src/winbind.c:52]: actually can't for two reasons.
NULL is returned as error if allocating context fail in the "normal" case.
Also NULL is used to figure out if we need to request a context explicitly in external_get_creds() which can be called both from a ASC/ISC call where a context is available and other calls where preallocated contexts are not available and therefore need to be set up on the fly.

**Comment by simo5** [comment]
@kvv81 ok updated patch to add a build time option to flip tls ctx on or off by default
The env var variable can still be used (but now required to be defined to either '0' or '1' to turn on or off tls storage usage '1' turns it on, '0' turns it off.

**Comment by simo5** [comment]
> Little typo: calso

@kvv81 Ok corrected it locally, will push once I have more feedback (and any other change needed due to that).

File: src/gss_creds.c [hunk 6]
```diff
@@ -520,7 +521,7 @@ uint32_t gssntlm_acquire_cred(uint32_t *minor_status,
                               gss_OID_set *actual_mechs,
                               uint32_t *time_rec)
 {
-    return gssntlm_acquire_cred_from(minor_status,
+    return gssntlm_acquire_cred_from(minor_status, NULL,
```
Comment (frozencemetery) [inline src/gss_creds.c:524]: Nit: adding these on the same line doesn't fit the local style of one argument per line.

**Review by frozencemetery — COMMENTED** [review]
I think I don't fully understand why this is a compile-time option.  If there's an env var, then there's little value in that, right?  It's just more confusing.

File: src/gss_creds.c (hunk 6 above)
Comment (simo5) [reply src/gss_creds.c:524]: fixed locally, will push later when full reviews are in

**Comment by kvv81** [comment]
> I think I don't fully understand why this is a compile-time option. If there's an env var, then there's little value in that, right? It's just more confusing.

Compile-time option gives default behaviour (if env var is absent).
And if env var is set, it overrides default behaviour.

We need compile-time (default) option because we have troubles with passing env variables from non-privileged parent to sudo-like process.

Both options are important - compile-time for us and env var for general purpose with no restrictions.

**Comment by frozencemetery** [comment]
> We need compile-time (default) option because we have troubles with passing env variables from non-privileged parent to sudo-like process.

But that should only be a problem if `secure_getenv()` were being used, and it's not - the PR uses `getenv()`.

**Comment by simo5** [comment]
> > We need compile-time (default) option because we have troubles with passing env variables from non-privileged parent to sudo-like process.
> 
> But that should only be a problem if `secure_getenv()` were being used, and it's not - the PR uses `getenv()`.

The problem is that many tools intentionally clean (or recreate from scratch) the environ before executing, so the env variable is simply not there at all.

**Comment by kvv81** [comment]
> > We need compile-time (default) option because we have troubles with passing env variables from non-privileged parent to sudo-like process.
> 
> But that should only be a problem if `secure_getenv()` were being used, and it's not - the PR uses `getenv()`.

Yes, you are right. But since you are already using it in GSSNTLMSSP for other variables (see GSSNTLMSSP_DEBUG - called with secure_getenv) so we can't utilize env vars as general solution for all needs.

**Comment by kvv81** [comment]
Sorry for late feedback...

We have tested this patch on stress-test (hardware SMB server + LoadDynamics networking system) to run SMB load.
The scenario was to run SMB NEGOTIATE+SESSION_SETUP(NTLM) + LOGOFF in a loop for multiple connections.
SMB server was using gssntlmssp, and winbind was a backend that connected to ActiveDirectory DC.
Single SMB server (with once instance of winbind) was used, winbind was calling to one DC.

Additionally I have configured those parameters for winbind to achieve better performance:

```
winbind max domain connections=40
log level=0 winbind:0
```

For a single connection we have got about 7ms response time for the 2nd SESSION_SETUP request (NTLM auth consists of two SESSION_SETUP requests, the 1st is processed very fast, and the 2nd is the actual talk with DC). In this case there were no races for resources, winbind and DC was fully dedicated to handle requests of only one client. This is a bare minimum of response time (having our hardware and networking).

Single connection:
image

For unfixed version (where NTLM authentication requests are just serialized in wbclient) we have got too large response time for 100 parallel connections because clients were waiting for each other (each auth takes 7ms, the rest of time the client was waiting in request queue). Total response time was about linear to amount of parallel clients.

100 parallel connections, no fix:
image

After the fix for gssntlmssp was applied, we have got much better results (the same scenario of 100 parallel connections).
The performance boost was significant, we have seen DC saturation (95% CPU load on ActiveDirectory side).

100 parallel connections, fixed version:
image

From winbind/gssntlmssp side, we haven't observed any problems under the stress (no auth errors).
Bottom line - the fix is very useful for high-performance systems, where many clients can work in parallel.

**Comment by simo5** [comment]
@kvv81 thank you so much for the exhaustive testing

sounds like this is a net improvement, so merging.
